### ISSUE TITLE
Compatibility with 2.3

### ISSUE THREAD

**Issue by gaiterjones** [mooore-digital/magento2-module-elasticsearch-relevance#1]
As far as I can see your plugin should work with the latest 2.3.x elasticsearch modules but I'm not seeing any changes to my search results when I raise the min_score value. When I manually do an elasticsearch search using 

`curl -XGET 'elasticsearch:9200/magento2_product_2_v1/_search?q=name:mysearch&pretty'`

I can see max result scores at around 25.0 if I add 25 as the min_score value I still see the same search results. 

I have checked the plugin code is definitely being executed.

**Comment by tdgroot** [comment]
Hi @gaiterjones,

Thank you for filing this issue.

The scoring model is [quite complicated](https://www.elastic.co/guide/en/elasticsearch/guide/current/practical-scoring-function.html) and perhaps the statement that 25 should be the minimum `min_score` might be incorrect. 

Please try setting lower `min_score` values. It really depends on the amount of filters, attributes, etc your shop has.

**Comment by gaiterjones** [comment]
Thanks, I will do some more testing. Just to confirm my original question this should still work with 2.3.4 and Elasticsearch 6.x?

**Comment by tdgroot** [comment]
This should still work on Magento 2.3.4 and ES 6.x. 

My previous answer was a bit vague about suggesting what to do. Let me clear it up a little.

There's a big difference between finding the right score using raw ES queries and using Magento search queries. The queries impact the scoring result, so you should really check what values are normal within the Magento search results. 

Also, try increasing it instead of lowering it. Try setting it to `100`. Does that limit too much? Try setting it to `50`, etc. 

You'll likely find the right `min_score` using that method and you'll have nice and tidy search results!

Please let me know about your findings.

**Comment by gaiterjones** [comment]
So it turned out the ES query result scores were just a lot higher that I had thought! I added a results debugger plugin to log the query results coming from ES and it showed I needed a `min score` value of between 400 and 500.

The debugger plugin looks like this and might be useful to include for other users having the same issue.

```
    public function beforeCreate(\Magento\Elasticsearch\SearchAdapter\ResponseFactory $subject, $result)
    {
       if($this->debug)
       {
           if(!is_array($result) || empty($result)) return false;

           $scores=array();
           foreach ($result['documents'] as $rawDocument) {
               $this->logger->debug('ELASTIC_SEARCH_QUERY_RESULTS_DEBUG',$rawDocument);
               array_push($scores,$rawDocument['_score']);
           }

           $this->logger->debug('ELASTIC_SEARCH_QUERY_RESULTS_DEBUG',array(
               'results' => count($rawDocument),
               'min_relevance_score' => $this->configuration->getMinScore(),
               'min_score' => min($scores),
               'max_score' => max($scores)
            )
           );
       }
    }
```

You can monitor the debug using `tail -F debug.log | grep ELASTIC_SEARCH_QUERY_RESULTS_DEBUG`

`[2020-05-27 09:47:44] main.DEBUG: ELASTIC_SEARCH_QUERY_RESULTS_DEBUG {"_index":"magento2_product_1_v586","_type":"document","_id":"124","_score":478.73044} []
[2020-05-27 09:47:44] main.DEBUG: ELASTIC_SEARCH_QUERY_RESULTS_DEBUG {"_index":"magento2_product_1_v586","_type":"document","_id":"44","_score":400.47678} []
[2020-05-27 09:47:44] main.DEBUG: ELASTIC_SEARCH_QUERY_RESULTS_DEBUG {"results":4,"min_relevance_score":375,"min_score":400.47678,"max_score":478.73044} []
`

My search results now look a lot better.

I think this min_score relevance setting should be a core feature of M2 as well as other ES search tweaks such as Ngram settings to help improve search results.  

I also noticed you are declaring the plugin twice in di.xml which I suspect is an error.

Thanks very much for making this info available!

**Comment by tdgroot** [comment]
Great to see you have found the right scores for your shop. 

Thank you for sharing your debugging process to find the right value!

> I think this min_score relevance setting should be a core feature of M2 as well as other ES search tweaks such as Ngram settings to help improve search results.

I agree that this should be a feature within Magento itself, but I'm really curious how they would implement it in a way that a user could easily find out what the right values are. 

> I also noticed you are declaring the plugin twice in di.xml which I suspect is an error.

Nice catch! I don't think it could do any harm in the case of this plugin, but it indeed is an error. I fixed that and released version `1.0.1`. 

> Thanks very much for making this info available!

You're welcome and thank you too for using the module and giving feedback!
